nni/retiarii/oneshot/pytorch/differentiable.py

```diff
@@ -0,0 +1,387 @@
+                for j in range(len(self.ops)):
+                    self.alpha.grad[i] += binary_grads[j] * probs[j] * (int(i == j) - probs[i])
+
+    def export(self):
+        return torch.argmax(self.alpha).item()
+
+    def export_prob(self):
+        return F.softmax(self.alpha, dim=-1)
+
+
+class ProxylessInputChoice(nn.Module):
+    def __init__(self, input_choice):
+        super().__init__()
+        self.ops = nn.ModuleList(input_choice)
+        self.alpha = nn.Parameter(torch.randn(input_choice.n_candidates) * 1E-3)
+        self._binary_gates = nn.Parameter(torch.randn(input_choice.n_candidates) * 1E-3)
+        self.sampled = None
+
+    def forward(self, inputs):
+        if self.training:
+            def run_function(active_sample):
+                return lambda x: x[active_sample]
+
+            def backward_function(binary_gates):
+                def backward(_x, _output, grad_output):
+                    binary_grads = torch.zeros_like(binary_gates.data)
+                    with torch.no_grad():
+                        for k in range(self.num_input_candidates):
+                            out_k = _x[k].data
+                            grad_k = torch.sum(out_k * grad_output)
+                            binary_grads[k] = grad_k
+                    return binary_grads
+                return backward
+
+            inputs = torch.stack(inputs, 0)
+            return _ArchGradientFunction.apply(
+                inputs, self._binary_gates, run_function(self.sampled),
+                backward_function(self._binary_gates)
+            )
+
+        return super().forward(inputs)
+
+    def resample(self, sample=None):
+        if sample is None:
+            probs = F.softmax(self.alpha, dim=-1)
+            sample = torch.multinomial(probs, 1)[0].item()
+        self.sampled = sample
+        with torch.no_grad():
+            self._binary_gates.zero_()
+            self._binary_gates.grad = torch.zeros_like(self._binary_gates.data)
+            self._binary_gates.data[sample] = 1.0
+        return self.sampled
+
+    def finalize_grad(self):
+        binary_grads = self._binary_gates.grad
+        with torch.no_grad():
+            if self.alpha.grad is None:
+                self.alpha.grad = torch.zeros_like(self.alpha.data)
+            probs = F.softmax(self.alpha, dim=-1)
+            for i in range(self.num_input_candidates):
+                for j in range(self.num_input_candidates):
+                    self.alpha.grad[i] += binary_grads[j] * probs[j] * (int(i == j) - probs[i])
+
+
+class ProxylessModule(DartsModule):
+    """
+    The Proxyless Module. This is a darts-based method. What it differs from darts is that it resample
+    the architecture according to alphas to select only one path a time to reduce memory consumption.
+    The Proxyless Model should be trained with ParallelTraiValDataloader in nn.retiarii.oneshot.pytorch.utils.
+    See base class for more attributes.
```

Comment:
Attributes from base class will be automatically rendered. No need to explicitly specify this.